diff --git a/CMakeLists.txt b/CMakeLists.txt
index 202cc52..b5c2bee 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -40,6 +40,8 @@ SET(UNITTEST_SOURCE_FILES
     t/stream-concurrency.c
     t/test.c)

+SET(CMAKE_POSITION_INDEPENDENT_CODE ON)
+
 ADD_LIBRARY(quicly ${QUICLY_LIBRARY_FILES})

 ADD_EXECUTABLE(cli ${PICOTLS_OPENSSL_FILES} src/cli.c)
@@ -50,6 +52,16 @@ TARGET_LINK_LIBRARIES(test.t quicly ${OPENSSL_LIBRARIES} ${CMAKE_DL_LIBS})

 ADD_EXECUTABLE(udpfw t/udpfw.c)

+INSTALL (
+    DIRECTORY ${CMAKE_SOURCE_DIR}/include/
+    DESTINATION include
+    FILES_MATCHING PATTERN "*.h")
+
+INSTALL(TARGETS quicly
+        RUNTIME DESTINATION bin
+        LIBRARY DESTINATION lib
+        ARCHIVE DESTINATION lib)
+
 ADD_CUSTOM_TARGET(check env BINARY_DIR=${CMAKE_CURRENT_BINARY_DIR} prove --exec "sh -c" -v ${CMAKE_CURRENT_BINARY_DIR}/*.t t/*.t
     WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
     DEPENDS cli test.t)
Submodule deps/picotls contains untracked content
Submodule deps/picotls contains modified content
diff --git a/deps/picotls/CMakeLists.txt b/deps/picotls/CMakeLists.txt
index cda6aad..62b23b7 100644
--- a/deps/picotls/CMakeLists.txt
+++ b/deps/picotls/CMakeLists.txt
@@ -5,6 +5,7 @@ PROJECT(picotls)

 FIND_PACKAGE(PkgConfig REQUIRED)

+SET(CMAKE_POSITION_INDEPENDENT_CODE ON)
 SET(CMAKE_C_FLAGS "-std=c99 -Wall -O2 -g ${CC_WARNING_FLAGS} ${CMAKE_C_FLAGS}")
 INCLUDE_DIRECTORIES(${OPENSSL_INCLUDE_DIR} deps/cifra/src/ext deps/cifra/src deps/micro-ecc deps/picotest include)
 SET(MINICRYPTO_LIBRARY_FILES
@@ -97,3 +98,12 @@ IF (BUILD_FUZZER)


 ENDIF()
+
+INSTALL (DIRECTORY ${CMAKE_SOURCE_DIR}/include/
+    DESTINATION include
+    FILES_MATCHING PATTERN "*.h")
+
+INSTALL(TARGETS picotls-core picotls-openssl
+        RUNTIME DESTINATION bin
+        LIBRARY DESTINATION lib
+        ARCHIVE DESTINATION lib)
