 drivers/net/ena/ena_ethdev.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/net/ena/ena_ethdev.c b/drivers/net/ena/ena_ethdev.c
index 7d4a3b225..26d875412 100644
--- a/drivers/net/ena/ena_ethdev.c
+++ b/drivers/net/ena/ena_ethdev.c
@@ -1599,6 +1599,7 @@ ena_set_queues_placement_policy(struct ena_adapter *adapter,
 				struct ena_admin_feature_llq_desc *llq,
 				struct ena_llq_configurations *llq_default_configurations)
 {
+#if 0
 	int rc;
 	u32 llq_feature_mask;
 
@@ -1630,6 +1631,10 @@ ena_set_queues_placement_policy(struct ena_adapter *adapter,
 	}
 
 	ena_dev->mem_bar = adapter->dev_mem_base;
+#else
+    RTE_LOG(INFO, PMD, "Forcing host mode policy (disabling LLQ for performance).\n");
+    ena_dev->tx_mem_queue_type = ENA_ADMIN_PLACEMENT_POLICY_HOST;
+#endif
 
 	return 0;
 }
