diff --git a/scapy/layers/l2.py b/scapy/layers/l2.py
index eee3fd68..29696a6c 100644
--- a/scapy/layers/l2.py
+++ b/scapy/layers/l2.py
@@ -424,6 +424,19 @@ def l2_register_l3_arp(l2, l3):
 conf.neighbor.register_l3(Ether, ARP, l2_register_l3_arp)


+class ERSPAN(Packet):
+    name = "ERSPAN"
+    fields_desc = [ BitField("ver",0,4),
+                    BitField("vlan",0,12),
+                    BitField("cos",0,3),
+                    BitField("en",0,2),
+                    BitField("t",0,1),
+                    BitField("session_id",0,10),
+                    BitField("reserved",0,12),
+                    BitField("index",0,20),
+                    ]
+
+
 class GRErouting(Packet):
     name = "GRE routing information"
     fields_desc = [ShortField("address_family", 0),
@@ -537,6 +550,7 @@ bind_layers(Ether, Dot1AD, type=0x88a8)
 bind_layers(Dot1AD, Dot1AD, type=0x88a8)
 bind_layers(Dot1AD, Dot1Q, type=0x8100)
 bind_layers(Dot1Q, Dot1AD, type=0x88a8)
+bind_layers(ERSPAN, Ether)
 bind_layers(Ether, Ether, type=1)
 bind_layers(Ether, ARP, type=2054)
 bind_layers(CookedLinux, LLC, proto=122)
@@ -549,6 +563,7 @@ bind_layers(GRE, Dot1Q, proto=33024)
 bind_layers(GRE, Dot1AD, type=0x88a8)
 bind_layers(GRE, Ether, proto=0x6558)
 bind_layers(GRE, ARP, proto=2054)
+bind_layers(GRE, ERSPAN, proto=0x88be, seqnum_present=1)
 bind_layers(GRE, GRErouting, {"routing_present": 1})
 bind_layers(GRErouting, conf.raw_layer, {"address_family": 0, "SRE_len": 0})
 bind_layers(GRErouting, GRErouting, {})
