diff --git a/scapy/layers/vxlan.py b/scapy/layers/vxlan.py
index 03c3a5a..ee30f28 100644
--- a/scapy/layers/vxlan.py
+++ b/scapy/layers/vxlan.py
@@ -69,6 +69,7 @@ bind_layers(UDP, VXLAN, dport=4789)  # RFC standard vxlan port
 bind_layers(UDP, VXLAN, dport=4790)  # RFC standard vxlan-gpe port
 bind_layers(UDP, VXLAN, dport=6633)  # New IANA assigned port for use with NSH
 bind_layers(UDP, VXLAN, dport=8472)  # Linux implementation port
+bind_layers(UDP, VXLAN, dport=48879) # iVXLAN port
 bind_layers(UDP, VXLAN, sport=4789)
 bind_layers(UDP, VXLAN, sport=4790)
 bind_layers(UDP, VXLAN, sport=6633)
