From 8ea104c5c772c176756ce5432d1d71f90c92b3c0 Mon Sep 17 00:00:00 2001
From: fenglei <1579628578@qq.com>
Date: Fri, 17 Jan 2025 16:59:02 +0800
Subject: [PATCH] xdp-dump: disable xdp-dump

Signed-off-by: fenglei <1579628578@qq.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 6a7f575..b445037 100644
--- a/Makefile
+++ b/Makefile
@@ -16,7 +16,7 @@ endif
 include version.mk
 include config.mk
 
-UTILS := xdp-filter xdp-loader xdp-dump
+UTILS := xdp-filter xdp-loader
 
 ifneq ($(BPFTOOL),)
 UTILS += xdp-bench xdp-forward xdp-monitor xdp-trafficgen
-- 
2.43.0

