From 939eb3d245e794679abc3c90be035e073a756b78 Mon Sep 17 00:00:00 2001
From: Fan Zhang <roy.fan.zhang@intel.com>
Date: Thu, 13 Jan 2022 09:44:46 +0000
Subject: [PATCH] common/qat: add gen4 b support

Signed-off-by: Fan Zhang <roy.fan.zhang@intel.com>
---
 drivers/common/qat/qat_device.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/common/qat/qat_device.c b/drivers/common/qat/qat_device.c
index 437996f2e8..1f870d689a 100644
--- a/drivers/common/qat/qat_device.c
+++ b/drivers/common/qat/qat_device.c
@@ -47,6 +47,9 @@ static const struct rte_pci_id pci_id_qat_map[] = {
 		{
 			RTE_PCI_DEVICE(0x8086, 0x4941),
 		},
+		{
+			RTE_PCI_DEVICE(0x8086, 0x4943),
+		},
 		{.device_id = 0},
 };
 
@@ -192,6 +195,7 @@ qat_pci_device_allocate(struct rte_pci_device *pci_dev,
 		qat_dev_gen = QAT_GEN3;
 		break;
 	case 0x4941:
+	case 0x4943:
 		qat_dev_gen = QAT_GEN4;
 		break;
 	default:
-- 
2.25.1

